### ISSUE TITLE
linfa-kernel fails to compile

### ISSUE THREAD

**Issue by AVS-Origami** [rust-ml/linfa#307]
When I try to build the wine quality example for linfa-svm, I get the following error:

```
   Compiling linfa-kernel v0.6.1
error[E0599]: no method named `mul` found for reference `&CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>>` in the current scope
  --> C:\Users\[username]\.cargo\registry\src\index.crates.io-6f17d22bba15001f\linfa-kernel-0.6.1\src\inner.rs:64:14
   |
64 |         self.mul(rhs)
   |              ^^^ method not found in `&CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>>`

error[E0599]: no method named `mul` found for reference `&CsMatBase<F, usize, &'a [usize], &'a [usize], &'a [F]>` in the current scope
   --> C:\Users\[username]\.cargo\registry\src\index.crates.io-6f17d22bba15001f\linfa-kernel-0.6.1\src\inner.rs:109:14
    |
109 |         self.mul(rhs)
    |              ^^^ method not found in `&CsMatBase<F, usize, &[usize], &[usize], &[F]>`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `linfa-kernel` (lib) due to 2 previous errors
```

I am on windows 10 with rustc version 1.70.0. I also tried version 0.5.1 which gives me the same error, and it also happens when I try it on my Arch Linux machine with rustc version 1.68.0.

I looked around in the source files, and it appears that this problem is related to the trait binding of the generic `F`:

```rs
impl<F: linfa::Float> Inner for CsMat<F>
```

and 

```rs
impl<'a, F: linfa::Float> Inner for CsMatView<'a, F>
```

Any help would be appreciated.

**Comment by xosxos** [comment]
I have the same issue. Seems like `CsMat` which is imported from `sprs` cannot implement `std::ops:Mul` for some reason. Multiplication of a `CsMat` with an `ArrayView2` is the issue which was  handled by the `std::ops::Mul` trait previously from what I can see. But for some reason this is now broken.

**Comment by AVS-Origami** [comment]
Yes. I tried creating a new `CsMat` in a test crate, and I was able to call `mul()` on it. So I think the issue is related to `CsMat` not liking `linfa::Float` as the generic.

**Comment by AVS-Origami** [comment]
I changed the start of the `impl` block to this:
```rs
impl<F: Float> Inner for CsMat<F> where 
    CsMat<F>: Mul {
        ...
}
```
which results in several errors that look like this:
```
error[E0277]: cannot multiply `CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>>` by `CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>>`
   --> algorithms/linfa-kernel/src/lib.rs:196:20
    |
196 | impl<'a, F: Float> Kernel<F> {
    |                    ^^^^^^^^^ no implementation for `CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>> * CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>>`
    |
    = help: the trait `Mul` is not implemented for `CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>>`
    = help: the following other types implement trait `Mul<Rhs>`:
              <&'a CsMatBase<N, I, IpS, IS, DS, Iptr> as Mul<&'b ndarray::ArrayBase<DS2, Dim<[usize; 1]>>>>
              <&'a CsMatBase<N, I, IpS, IS, DS, Iptr> as Mul<&'b ndarray::ArrayBase<DS2, Dim<[usize; 2]>>>>
              <&'a CsMatBase<N, I, IpS1, IS1, DS1, Iptr> as Mul<&'b CsMatBase<N, I, IpS2, IS2, DS2, Iptr>>>
              <&'a CsMatBase<N, I, IpS1, IS1, DS1, Iptr> as Mul<&'b CsVecBase<IS2, DS2, N, I>>>
              <&'a CsMatBase<f32, I, IpStorage, IStorage, DStorage, Iptr> as Mul<f32>>
              <&'a CsMatBase<f64, I, IpStorage, IStorage, DStorage, Iptr> as Mul<f64>>
              <&'a CsMatBase<i16, I, IpStorage, IStorage, DStorage, Iptr> as Mul<i16>>
              <&'a CsMatBase<i32, I, IpStorage, IStorage, DStorage, Iptr> as Mul<i32>>
            and 8 others
note: required for `CsMatBase<F, usize, Vec<usize>, Vec<usize>, Vec<F>>` to implement `Inner`
   --> algorithms/linfa-kernel/src/inner.rs:60:16
    |
60  | impl<F: Float> Inner for CsMat<F> where 
    |                ^^^^^     ^^^^^^^^
61  |     CsMat<F>: Mul {
    |               --- unsatisfied trait bound introduced here
```

**Comment by bytesnake** [comment]
Could be related to https://github.com/sparsemat/sprs/issues/278

**Comment by bytesnake** [comment]
This is a regression introduced with the `sprs::MulAcc` trait in 0.11 https://github.com/sparsemat/sprs/blob/master/changelog.rst

**Comment by AVS-Origami** [comment]
It works now. Thanks!

**Comment by jeff-hykin** [comment]
I still get this issue. What version is the fix on?
<img width="1274" alt="Screen Shot 2023-07-06 at 11 40 34 AM" src="https://github.com/rust-ml/linfa/assets/17692058/9d70b139-fac9-4805-97dd-d7edb60cdf74">

**Comment by jeff-hykin** [comment]
Thanks! @AVS-Origami

**Comment by clime** [comment]
I;ve just opened https://github.com/sparsemat/sprs/issues/330

**Comment by cgubbin** [comment]
I'm having this issue again today on linfa 0.7.0.

Thanks to the above I resolved by pinning `sprs = "=0.11.1"` and everything is compiling again. Would you accept a pull request to pin the `sprs` version?

**Comment by guaraqe** [comment]
I am seeing this again:

  --> /home/juan/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/linfa-kernel-0.8.0/src/inner.rs:64:18
   |
64 |         self.mul(rhs)
   |              --- ^^^ no implementation for `&CsMatBase<F, usize, ..., ..., ...> * &ArrayBase<ViewRepr<&F>, ...>`
   |              |
   |              required by a bound introduced by this call
   |
